### ISSUE TITLE
Train

### ISSUE THREAD

**Issue by yohehe** [yohehe/pytorch_temp#4]

**Comment by yohehe** [comment]
``
# Train Model
from tqdm import tqdm


def train_model(net, dataloaders_dict, criterion, optimizer, num_epochs):

    # 初期設定
    # GPUが使えるかを確認
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    print("使用デバイス：", device)

    # ネットワークをGPUへ
    net.to(device)

    # ネットワークがある程度固定であれば、高速化させる
    torch.backends.cudnn.benchmark = True
    
    
    #lossなどを格納するlistを設定
    loss_list=[]
    acc_list=[]
    
    f1_list=[]
    recall_list=[]
    precision_list=[]
    
    
    #validationデータの評価時に格納するlist
    val_loss_list=[]
    val_acc_list=[]
    
    val_f1_list=[]
    val_recall_list=[]
    val_precision_list=[]
    

    # epochのループ
    for epoch in range(num_epochs):
        print('Epoch {}/{}'.format(epoch+1, num_epochs))
        print('-------------')

        # epochごとの訓練と検証のループ
        for phase in ['train', 'val']:
            if phase == 'train':
                net.train()  # モデルを訓練モードに
            else:
                net.eval()   # モデルを検証モードに

            epoch_loss = 0.0  # epochの損失和
            epoch_corrects = 0  # epochの正解数
            epoch_f1 = 0.
            epoch_recall=0.
            epoch_precision=0.

            # 未学習時の検証性能を確かめるため、epoch=0の訓練は省略
            if (epoch == 0) and (phase == 'train'):
                continue

            # データローダーからミニバッチを取り出すループ
            for inputs, labels in tqdm(dataloaders_dict[phase]):

                # GPUが使えるならGPUにデータを送る
                inputs = inputs.to(device)
                labels = labels.to(device)
          
                # optimizerを初期化
                optimizer.zero_grad()

                # 順伝搬（forward）計算
                with torch.set_grad_enabled(phase == 'train'):

                    outputs= net(inputs)                    
                    loss = criterion(outputs, labels)  # 損失を計算
                    
                    #batch_sizeのconfusion_matrixの平均を出力する
                    batch_f1,batch_precision,batch_recall=mean_accuracy(inputs=outputs,
                                                      labels=labels)
                    """
                    print("f1:{:.3f},precision:{:.3f},recall:{:.3f}"\
                          .format(batch_f1,batch_precision,batch_recall)
                         )
                    """
                    
                    # 訓練時はバックプロパゲーション
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()

                    # 結果の計算
                    epoch_loss += loss.item() * inputs.size(0)  # lossの合計を更新
                    
                    # 正解数の合計を更新
                    #epoch_corrects += torch.sum(preds == labels.data)
                    #confusion_matrixの設定出力を更新
                    
                    epoch_f1 += batch_f1
                    epoch_recall += batch_recall
                    epoch_precision += batch_precision
            
            # epochごとのlossと正解率を表示
            epoch_loss = epoch_loss / len(dataloaders_dict[phase].dataset)
            #epoch_acc = epoch_corrects.double() / len(dataloaders_dict[phase].dataset)
            
            epoch_f1_data= epoch_f1 / len(dataloaders_dict[phase].dataset)
            epoch_recall_data = epoch_recall / len(dataloaders_dict[phase].dataset)
            epoch_precision_data = epoch_precision / len(dataloaders_dict[phase].dataset)
            
            if phase=="train":
                loss_list.append(epoch_loss)
                #acc_list.append(epoch_acc)
                
                """
                f1_list.append(epoch_f1_data)
                recall_list.append(epoch_recall_data)
                precision_list.append(epoch_precision_data)
                """
            else:
                val_loss_list.append(epoch_loss)
                #val_acc_list.append(epoch_acc)
                """
                val_f1_list.append(epoch_f1_data)
                val_recall_list.append(epoch_recall_data)
                val_precision_list.append(epoch_precision_data)
                """
           
            print('{} Loss: {:.4f} ,, F1:{:.4f}, Recall:{:.4f}, Precision:{:.4f}'.format(
                phase, epoch_loss, epoch_f1_data,epoch_recall_data,epoch_precision_data))
           
            
            #print('{} Loss: {:.4f} ,'.format(phase, epoch_loss,))
            
        """
        else:    
            #グラフ表示の設定:
            fig,(axL,axR)=plt.subplots(ncols=2,figsize=(20,5))
            # plot learning curve
            #plt.figure()
            #1回目の計算を取得しないようにしているため,-1を行う必要あり。
            axL.plot(loss_list,color='skyblue', label='loss')
            axL.plot(val_loss_list, color='orange', label='val_loss')
            axL.legend()
            axL.set_xlabel('epochs')
            axL.set_ylabel('loss')
            axL.grid(True)

            #plt.figure()
            #axR.plot(acc_list, color='skyblue', label='acc')
            #axR.plot(val_acc_list, color='orange', label='val_acc')
            #axR.legend()
            #axR.set_xlabel('epochs')
            #axR.set_ylabel('accuracy')
            axR.grid(True)
        """
``
